### PR TITLE
Add pinning scores

### LINKED ISSUE #2664: Add user option to pin plays to their profile
https://www.reddit.com/r/osugame/comments/835xwo/feature_idea_being_able_to_choose_what_plays/

Has been brought up before as well, wanted to bring it to github in case you weren't aware that it could be a neat little addition to optionally show off cool scores that aren't high pp plays.

### PR DESCRIPTION
**Pull request by nanaya**
Reordering will come later :tm: 

No design for score-top component and scores-show page :eyes: 

- [x] depends on #8433

Resolves #2664. Note though only best play is currently available for pinning and if pinned score is replaced by another score, it'll be gone.

### CODE REVIEW

File: resources/assets/lib/profile-page/play-detail-list.tsx [hunk 1]
```diff
@@ -24,6 +24,10 @@ const sectionMaps = {
     countKey: 'scores_first_count',
     translationKey: 'top_ranks.first',
   },
+  scoresPinned: {
+    countKey: 'scores_pinned_count',
```
Comment (notbakaneko) [inline resources/assets/lib/profile-page/play-detail-list.tsx:28]: also need to update the count somehow after pin/unpin 🤔

**Review by notbakaneko — CHANGES_REQUESTED** [review]
first pass...

**Comment by notbakaneko** [comment]
> if pinned score is replaced by another score, it'll be gone.
what happens to the existing entry in the database?

**Comment by nanaya** [comment]
> what happens to the existing entry in the database?

to be cleaned up with some scheduled job task later™

File: resources/assets/lib/components/score-pin.tsx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+  constructor(props: Props) {
+    super(props);
+
+    makeObservable(this);
+  }
+
+  componentWillUnmount() {
+    this.onUpdateCallbacks = null;
```
Comment (notbakaneko) [inline resources/assets/lib/components/score-pin.tsx:43]: the active menu key still needs to be `null`ed; `PopupMenuPersistent` unmount seems to be the appropriate place for this, although `useEffect` seems to trigger too many times for each change 🤔

File: resources/assets/lib/core/user/score-pins.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import OsuCore from 'osu-core';
+
+export default class ScorePins {
+  @observable pins = new Map<string, boolean>();
+
+  constructor(private core: OsuCore) {
+    makeObservable(this);
+  }
+
+  apiPin(score: ScoreJson, toPin: boolean): JQuery.jqXHR<void> {
```
Comment (notbakaneko) [inline resources/assets/lib/core/user/score-pins.ts:16]: return type of of the `ajax` call should be casted to be not `<any>`

**Review by notbakaneko — CHANGES_REQUESTED** [review]
weighted score disappearing after reload but newly pinned score having weight seems strange
image

File: resources/assets/lib/core/user/score-pins.ts (hunk 3 above)
Comment (nanaya) [reply resources/assets/lib/core/user/score-pins.ts:16]: Doesn't seem to matter? 🤔

Comment (notbakaneko) [reply resources/assets/lib/core/user/score-pins.ts:16]: returning `jqXHR<any>` for a function typed `jqXHR<void>` is a lint warning

**Review by notbakaneko — APPROVED** [review]

**Comment by Loreos7** [comment]
Will it be available to translate?

**Comment by peppy** [comment]
yes, please wait until we sync localisations.

**Pull request merged**